File: src/strats_prob/aggregator.py
"""Trade result aggregation for probe system.

Groups trades by (open_day=date, open_hour, long_short) and computes
aggregate statistics for each group.
"""

import logging
from typing import Optional

import numpy as np

from src.strats_prob.strategy_def import ProbeTradeResult

logger = logging.getLogger(__name__)


def aggregate_trades(
    trades: list[ProbeTradeResult],
    strategy_id: int,
    symbol: str,
    timeframe: str,
    risk_profile: str,
    run_id: str,
    period_start,
    period_end,
) -> list[dict]:
    """Aggregate trade results into probe records grouped by dimensions.

    Groups by (open_day=date, open_hour, long_short) and computes:
    - num_trades
    - pnl_mean
    - pnl_std
    - max_drawdown (worst among all trades in group)
    - max_profit (best among all trades in group)

    Args:
        trades: List of ProbeTradeResult from the engine.
        strategy_id: FK to probe_strategies table.
        symbol: Ticker symbol.
        timeframe: Bar timeframe (e.g. "1Hour").
        risk_profile: Risk profile name (low/medium/high).
        run_id: Unique run identifier.
        period_start: Start of the evaluation period.
        period_end: End of the evaluation period.

    Returns:
        List of dicts ready for ProbeRepository.bulk_insert_results().
    """
    if not trades:
        logger.debug(
            "No trades to aggregate for strategy_id=%d, %s/%s/%s",
            strategy_id, symbol, timeframe, risk_profile,
        )
        return []

    # Group trades by dimensions
    groups: dict[tuple, list[ProbeTradeResult]] = {}
    for trade in trades:
        open_day = trade.entry_time.date()  # Full calendar date (YYYY-MM-DD)
        open_hour = trade.entry_time.hour
        long_short = trade.direction[:5]  # "long" or "short"

        key = (open_day, open_hour, long_short)
        if key not in groups:
            groups[key] = []
        groups[key].append(trade)

    # Compute aggregations per group
    records = []
    for (open_day, open_hour, long_short), group_trades in groups.items():
        pnls = [t.pnl_pct for t in group_trades]
        drawdowns = [t.max_drawdown_pct for t in group_trades]
        profits = [t.max_profit_pct for t in group_trades]

        pnl_arr = np.array(pnls)

        records.append({
            "run_id": run_id,
            "symbol": symbol.upper(),
            "strategy_id": strategy_id,
            "period_start": period_start,
            "period_end": period_end,
            "timeframe": timeframe,
            "risk_profile": risk_profile,
            "open_day": open_day,
            "open_hour": open_hour,
            "long_short": long_short,
            "num_trades": len(group_trades),
            "pnl_mean": float(np.mean(pnl_arr)),
            "pnl_std": float(np.std(pnl_arr)) if len(pnl_arr) > 1 else 0.0,
            "max_drawdown": float(max(drawdowns)) if drawdowns else 0.0,
            "max_profit": float(max(profits)) if profits else 0.0,
        })

    logger.debug(
        "Aggregated %d trades into %d groups for strategy_id=%d, %s/%s/%s",
        len(trades), len(records), strategy_id, symbol, timeframe, risk_profile,
    )
    return records

File: go-strats/pkg/persistence/aggregator.go
// Package persistence provides trade aggregation and database persistence
// for strategy probe results. It mirrors the Python aggregation logic in
// src/strats_prob/aggregator.py and persistence in src/strats_prob/runner.py.
package persistence

import (
	"math"
	"strings"
	"time"

	"github.com/algomatic/strats100/go-strats/pkg/types"
)

// GroupKey identifies a unique aggregation group matching the Python dimensions:
// (open_day=date, open_hour=0-23, long_short="long"|"short").
type GroupKey struct {
	OpenDay   time.Time // Calendar date only (YYYY-MM-DD), time part is zero
	OpenHour  int       // 0-23
	LongShort string    // "long" or "short"
}

// AggregatedResult holds the computed statistics for one group of trades.
// Maps directly to a row in the strategy_probe_results table.
type AggregatedResult struct {
	// Identification
	RunID       string
	Symbol      string
	StrategyID  int
	PeriodStart time.Time
	PeriodEnd   time.Time

	// Dimensions
	Timeframe   string
	RiskProfile string
	OpenDay     time.Time // Calendar date (YYYY-MM-DD)
	OpenHour    int
	LongShort   string

	// Aggregations
	NumTrades   int
	PnLMean     float64
	PnLStd      float64
	MaxDrawdown float64
	MaxProfit   float64
}

// TradeRecord holds the fields for one row in the strategy_probe_trades table.
type TradeRecord struct {
	ResultID           int64 // FK to strategy_probe_results.id (set after result insert)
	Ticker             string
	OpenTimestamp      time.Time
	CloseTimestamp     time.Time
	Direction          string
	OpenJustification  string
	CloseJustification string
	PnL                float64 // Currency P&L = pnl_pct * entry_price
	PnLPct             float64
	BarsHeld           int
	MaxDrawdown        float64
	MaxProfit          float64
	PnLStd             float64
}

// AggregateTrades groups trades by (open_day, open_hour, long_short) and
// computes per-group statistics. This matches the Python aggregate_trades()
// function in src/strats_prob/aggregator.py.
func AggregateTrades(
	trades []types.Trade,
	strategyID int,
	symbol, timeframe, riskProfile, runID string,
	periodStart, periodEnd time.Time,
) []AggregatedResult {
	if len(trades) == 0 {
		return nil
	}

	// Group trades by dimensions
	groups := make(map[GroupKey][]types.Trade)
	for _, t := range trades {
		key := GroupKey{
			OpenDay:   truncateToDate(t.EntryTime),
			OpenHour:  t.EntryTime.Hour(),
			LongShort: normalizeDirection(string(t.Direction)),
		}
		groups[key] = append(groups[key], t)
	}

	// Compute aggregations per group
	results := make([]AggregatedResult, 0, len(groups))
	for key, groupTrades := range groups {
		pnls := make([]float64, len(groupTrades))
		var maxDD, maxProfit float64

		for i, t := range groupTrades {
			pnls[i] = t.PnLPct
			if t.MaxDrawdownPct > maxDD {
				maxDD = t.MaxDrawdownPct
			}
			if t.MaxProfitPct > maxProfit {
				maxProfit = t.MaxProfitPct
			}
		}

		results = append(results, AggregatedResult{
			RunID:       runID,
			Symbol:      strings.ToUpper(symbol),
			StrategyID:  strategyID,
			PeriodStart: periodStart,
			PeriodEnd:   periodEnd,
			Timeframe:   timeframe,
			RiskProfile: riskProfile,
			OpenDay:     key.OpenDay,
			OpenHour:    key.OpenHour,
			LongShort:   key.LongShort,
			NumTrades:   len(groupTrades),
			PnLMean:     mean(pnls),
			PnLStd:      stddev(pnls),
			MaxDrawdown: maxDD,
			MaxProfit:   maxProfit,
		})
	}

	return results
}

// BuildTradeRecords converts engine trades to TradeRecord structs ready for
// DB insertion. The ResultID field is left as 0 and must be set after the
// corresponding AggregatedResult is inserted and its DB ID is known.
func BuildTradeRecords(trades []types.Trade, symbol string) []TradeRecord {
	records := make([]TradeRecord, len(trades))
	for i, t := range trades {
		records[i] = TradeRecord{
			Ticker:             strings.ToUpper(symbol),
			OpenTimestamp:      t.EntryTime,
			CloseTimestamp:     t.ExitTime,
			Direction:          normalizeDirection(string(t.Direction)),
			OpenJustification:  t.EntryJustification,
			CloseJustification: t.ExitJustification,
			PnL:                t.PnLPct * t.EntryPrice,
			PnLPct:             t.PnLPct,
			BarsHeld:           t.BarsHeld,
			MaxDrawdown:        t.MaxDrawdownPct,
			MaxProfit:          t.MaxProfitPct,
			PnLStd:             t.PnLStd,
		}
	}
	return records
}

// MapTradesToResults assigns each TradeRecord a ResultID by looking up its
// group key in the provided map. Trades that don't match any result are
// skipped (returned in the second slice).
func MapTradesToResults(
	tradeRecords []TradeRecord,
	trades []types.Trade,
	resultIDMap map[GroupKey]int64,
) (matched []TradeRecord, unmatched int) {
	matched = make([]TradeRecord, 0, len(tradeRecords))
	for i, tr := range tradeRecords {
		key := GroupKey{
			OpenDay:   truncateToDate(trades[i].EntryTime),
			OpenHour:  trades[i].EntryTime.Hour(),
			LongShort: tr.Direction,
		}
		if rid, ok := resultIDMap[key]; ok {
			tr.ResultID = rid
			matched = append(matched, tr)
		} else {
			unmatched++
		}
	}
	return matched, unmatched
}

// truncateToDate strips the time component, returning midnight UTC of that date.
func truncateToDate(t time.Time) time.Time {
	y, m, d := t.Date()
	return time.Date(y, m, d, 0, 0, 0, 0, time.UTC)
}

// normalizeDirection ensures the direction string is "long" or "short"
// (truncated to 5 chars to match Python's direction[:5]).
func normalizeDirection(d string) string {
	if len(d) > 5 {
		return d[:5]
	}
	return d
}

// mean computes the arithmetic mean of a float64 slice.
func mean(values []float64) float64 {
	if len(values) == 0 {
		return 0
	}
	sum := 0.0
	for _, v := range values {
		sum += v
	}
	return sum / float64(len(values))
}

// stddev computes the population standard deviation of a float64 slice.
// Returns 0 if fewer than 2 values (matching Python np.std with ddof=0,
// but returning 0 for single-element slices as in the aggregator).
func stddev(values []float64) float64 {
	n := len(values)
	if n <= 1 {
		return 0
	}
	m := mean(values)
	sumSq := 0.0
	for _, v := range values {
		d := v - m
		sumSq += d * d
	}
	return math.Sqrt(sumSq / float64(n))
}
